src/sentry/release_health/metrics.py

```diff
@@ -606,3 +611,131 @@ def extract_row_info(row: Mapping[str, Union[OrganizationId, str]]) -> ReleaseNa
+            Condition(Column("timestamp"), Op.GTE, start),
+        ]
+        if environment_values is not None:
+            conditions.append(Condition(Column(environment_key), Op.IN, environment_values))
+
+        def query_stats(end: datetime) -> CrashFreeBreakdown:
+            def _get_data(entity: EntityKey, metric_name: str) -> Tuple[int, int]:
+                total = 0
+                crashed = 0
+                metric_id = try_get_string_index(org_id, metric_name)
+                if metric_id is not None:
+                    where = conditions + [
+                        Condition(Column("metric_id"), Op.EQ, metric_id),
+                        Condition(Column("timestamp"), Op.LT, end),
+                    ]
+                    data = raw_snql_query(
+                        Query(
+                            dataset=Dataset.Metrics.value,
+                            match=Entity(entity.value),
+                            select=[Column("value")],
+                            where=where,
+                            groupby=[Column(status_key)],
+                        ),
+                        referrer="release_health.metrics.crash-free-breakdown.session",
+                    )["data"]
+                    for row in data:
+                        if row[status_key] == tag_value(org_id, "init"):
+                            total = int(row["value"])
+                        elif row[status_key] == tag_value(org_id, "crashed"):
+                            crashed = int(row["value"])
+
+                return total, crashed
+
+            sessions_total, sessions_crashed = _get_data(EntityKey.MetricsCounters, "session")
+            users_total, users_crashed = _get_data(EntityKey.MetricsSets, "user")
+
+            return {
+                "date": end,
+                "total_users": users_total,
+                "crash_free_users": 100 - users_crashed / float(users_total) * 100
+                if users_total
+                else None,
+                "total_sessions": sessions_total,
+                "crash_free_sessions": 100 - sessions_crashed / float(sessions_total) * 100
+                if sessions_total
+                else None,
+            }
+
+        return query_stats
+
+    def get_crash_free_breakdown(
+        self,
+        project_id: ProjectId,
+        release: ReleaseName,
+        start: datetime,
+        environments: Optional[Sequence[EnvironmentName]] = None,
+    ) -> Sequence[CrashFreeBreakdown]:
+
+        org_id = self._get_org_id([project_id])
+
+        now = datetime.now(pytz.utc)
+        query_fn = self._get_crash_free_breakdown_fn(
+            org_id, project_id, release, start, environments
+        )
+
+        last: Optional[datetime] = None
+        rv = []
+        for offset in (
+            timedelta(days=1),
+            timedelta(days=2),
+            timedelta(days=7),
+            timedelta(days=14),
```

Comment:
- What is the reason behind using a tuple here?
- Not a big fan of looping over the timedelta, so much repitition, why don't you just over the number of days and them call timedelta once outside of the loop

Reply:
tuples are cheaper to construct

Reply:
Sure but how expensive really is it to have a list of 4 items?

Reply:
I get the argument that lists are supposed to be the homogenous container and tuples the heterogenous but I don't buy it. Not a lot of code works this way and it seems like a style guideline that was invented long after tuples and lists were. For all practical purposes tuples are immutable lists as long as mypy doesn't get in your way about it.